### DISCUSSION TITLE
Support javascript debugging?

### DISCUSSION THREAD

**Discussion by luffyfly** [boa-dev/boa#2430] [Q&A] [answered]
I plan to use boa as the embedding script engine of the modding system in my game and it's necessary to support debugging.
Is there any plan to support debugging like Chrome devtools or just support debugging with Chrome devtools?

**Comment by jedel1043** [comment, marked as answer]
Yep! The spec defines a [debugger statement](https://tc39.es/ecma262/multipage/ecmascript-language-statements-and-declarations.html#sec-debugger-statement) which is precisely for use cases like yours.

Previously that was a bit impossible, since we were doing ast interpretation, but we recently moved to a VM architecture, which makes it a bit easier to implement pausing and resuming JS code. Though, that'll require some amount of API design.

If you have ideas about the API or some examples from which we can design ours, feel free to comment them here 😁

**Reply by luffyfly** [reply]
When I chose which engine to use, I compared `deno` and `boa` and I found `deno` supports `Chrome DevTools Protocol `, so. people who use `deno` can debug Javascript with Chrome devtools.
Maybe we can consider to implement `Chrome DevTools Protocol ` and people can debug JavaScript in boa with Chrome Devtools.

**Reply by luffyfly** [reply]
By the way, `Node.js` supports [debugging JavaScript  with Chrome DevTool](https://medium.com/@paul_irish/debugging-node-js-nightlies-with-chrome-devtools-7c4a1b95ae27), too.

**Reply by jasonwilliams** [reply]
Yes we certainly want to support this but I’m not sure when it will happen. There’s a bunch of other things that are our focus at the moment like making the engine more sound to use.
